File: packages/app/src/agent/toolCatalog.js
// @ts-check
import Ajv from "ajv";
import generatedToolCatalog from "./generated/generatedToolCatalog.json" with { type: "json" };
import generatedToolSchema from "./generated/generatedToolSchema.json" with { type: "json" };
import generatedActionCatalog from "./generated/generatedActionCatalog.json" with { type: "json" };
import { formatAjvErrors } from "./validationErrorFormatter.js";

// These generated artifacts are derived from agentToolInputs.d.ts and are the
// runtime source for tool descriptions, validation, and Responses API shapes.
const AjvClass = /** @type {any} */ (Ajv);

const ajv = new AjvClass({
    allErrors: true,
    strict: false,
    allowUnionTypes: true,
});

/** @type {Map<string, import("ajv").ValidateFunction>} */
const validatorsByToolName = new Map();

/**
 * @returns {import("./types.d.ts").AgentToolCatalogEntry[]}
 */
export function listAgentTools() {
    return generatedToolCatalog.map((entry) => ({
        ...entry,
        inputFields: entry.inputFields.map((field) => ({
            ...field,
        })),
    }));
}

/**
 * @returns {Array<{
 *     type: "function";
 *     name: string;
 *     description: string;
 *     parameters: Record<string, any>;
 *     strict: boolean;
 * }>}
 */
export function buildResponsesToolDefinitions() {
    return generatedToolCatalog.map((entry) => ({
        type: "function",
        name: entry.toolName,
        description: formatToolDescription(entry),
        parameters: getToolParameters(entry.inputType),
        strict: entry.strict !== false,
    }));
}

/**
 * Formats a tool-call rejection message using the generated tool catalog.
 *
 * @param {string} toolName
 * @param {string[]} errors
 * @returns {string}
 */
export function formatToolCallRejection(toolName, errors) {
    const tool = generatedToolCatalog.find(
        (entry) => entry.toolName === toolName
    );
    const action = generatedActionCatalog.find(
        (entry) => entry.actionType === toolName
    );

    const validationText =
        errors.length > 0
            ? "Validation errors:\n- " + errors.join("\n- ")
            : "Validation failed.";

    if (!tool) {
        if (action) {
            const exampleProgram = JSON.stringify(
                {
                    actions: [
                        {
                            actionType: toolName,
                            payload: action.examplePayload,
                        },
                    ],
                },
                null,
                2
            );

            return [
                "Tool call was incorrect and rejected. Correct it before trying again.",
                `${toolName} is an actionType, not a callable tool.`,
                "Use `submitIntentActions` and put that actionType inside `actions`.",
                "Example input:",
                exampleProgram,
                validationText,
            ].join("\n");
        }

        return (
            "Tool call was incorrect and rejected. Correct it before trying again. " +
            validationText
        );
    }

    const expectedFields = tool.inputFields
        .map((field) => `${field.name} (${field.type})`)
        .join(", ");
    const exampleInput = JSON.stringify(tool.exampleInput, null, 2);

    return [
        "Tool call was incorrect and rejected. Correct it before trying again.",
        `${tool.toolName} expects ${expectedFields}.`,
        "Example input:",
        exampleInput,
        validationText,
    ].join("\n");
}

/**
 * @param {string} toolName
 * @param {unknown} toolArguments
 * @returns {import("./types.d.ts").ShapeValidationResult}
 */
export function validateToolArgumentsShape(toolName, toolArguments) {
    const validator = getToolArgumentsValidator(toolName);
    if (!validator) {
        return {
            ok: false,
            errors: ["Unsupported agent tool " + toolName + "."],
        };
    }

    if (validator(toolArguments)) {
        return {
            ok: true,
            errors: [],
        };
    }

    return {
        ok: false,
        errors: formatAjvErrors("$", validator.errors),
    };
}

/**
 * @param {string} inputType
 * @returns {Record<string, any>}
 */
function getToolParameters(inputType) {
    const schemaDefinitions = /** @type {Record<string, any>} */ (
        generatedToolSchema.definitions ?? {}
    );
    const schema = schemaDefinitions[inputType];
    if (!schema) {
        throw new Error(
            "Missing generated schema for tool input " + inputType + "."
        );
    }

    return projectToolSchema(schema, schemaDefinitions);
}

/**
 * @param {import("./types.d.ts").AgentToolCatalogEntry} entry
 * @returns {string}
 */
function formatToolDescription(entry) {
    return entry.description;
}

/**
 * @param {string} toolName
 * @returns {import("ajv").ValidateFunction | undefined}
 */
function getToolArgumentsValidator(toolName) {
    const cachedValidator = validatorsByToolName.get(toolName);
    if (cachedValidator) {
        return cachedValidator;
    }

    const tool = generatedToolCatalog.find(
        (entry) => entry.toolName === toolName
    );
    if (!tool) {
        return undefined;
    }

    const schemaDefinitions = /** @type {Record<string, any>} */ (
        generatedToolSchema.definitions ?? {}
    );
    const schema = schemaDefinitions[tool.inputType];
    if (!schema) {
        throw new Error(
            "Missing generated schema for tool input " + tool.inputType + "."
        );
    }

    const validator = ajv.compile(getSchemaWrapper(schema));
    validatorsByToolName.set(toolName, validator);
    return validator;
}

/**
 * @param {Record<string, any>} schema
 * @param {Record<string, any>} definitions
 * @param {Set<string>} [excludedDefinitionNames]
 * @param {Set<string>} [visited]
 * @returns {Record<string, any>}
 */
function projectToolSchema(
    schema,
    definitions,
    excludedDefinitionNames = new Set(["AgentIntentBatchStep"]),
    visited = new Set()
) {
    if (schema === null || typeof schema !== "object") {
        return schema;
    }

    if (Array.isArray(schema)) {
        return schema.map((item) =>
            projectToolSchema(
                item,
                definitions,
                excludedDefinitionNames,
                visited
            )
        );
    }

    const ref = /** @type {{ $ref?: unknown }} */ (schema).$ref;
    if (typeof ref === "string") {
        const refName = ref.replace("#/definitions/", "");
        if (excludedDefinitionNames.has(refName) || visited.has(refName)) {
            return { type: "object" };
        }

        const refSchema = definitions[refName];
        if (!refSchema) {
            return { type: "object" };
        }

        visited.add(refName);
        const projected = projectToolSchema(
            refSchema,
            definitions,
            excludedDefinitionNames,
            visited
        );
        visited.delete(refName);
        return projected;
    }

    /** @type {Record<string, any>} */
    const projected = {};
    for (const [key, value] of Object.entries(schema)) {
        if (key === "definitions" || key === "$schema") {
            continue;
        }

        if (
            key === "properties" &&
            value &&
            typeof value === "object" &&
            !Array.isArray(value)
        ) {
            projected.properties = Object.fromEntries(
                Object.entries(value).map(([propertyName, propertySchema]) => [
                    propertyName,
                    projectToolSchema(
                        propertySchema,
                        definitions,
                        excludedDefinitionNames,
                        visited
                    ),
                ])
            );
            continue;
        }

        if (
            key === "items" ||
            key === "not" ||
            key === "if" ||
            key === "then" ||
            key === "else"
        ) {
            projected[key] = projectToolSchema(
                value,
                definitions,
                excludedDefinitionNames,
                visited
            );
            continue;
        }

        if (
            (key === "anyOf" || key === "allOf" || key === "oneOf") &&
            Array.isArray(value)
        ) {
            projected[key] = value.map((item) =>
                projectToolSchema(
                    item,
                    definitions,
                    excludedDefinitionNames,
                    visited
                )
            );
            continue;
        }

        if (
            key === "additionalProperties" &&
            value &&
            typeof value === "object" &&
            !Array.isArray(value)
        ) {
            projected[key] = projectToolSchema(
                value,
                definitions,
                excludedDefinitionNames,
                visited
            );
            continue;
        }

        projected[key] = value;
    }

    if (projected.type === "object" && projected.properties === undefined) {
        projected.properties = {};
    }

    return projected;
}

/**
 * @param {Record<string, any>} schema
 * @returns {Record<string, any>}
 */
function getSchemaWrapper(schema) {
    return {
        $schema: generatedToolSchema.$schema,
        definitions: generatedToolSchema.definitions,
        ...schema,
    };
}

File: utils/agent_server/app/tool_catalog.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

_REPO_ROOT = Path(__file__).resolve().parents[3]
_GENERATED_AGENT_DIR = _REPO_ROOT / "packages/app/src/agent/generated"
_TOOL_CATALOG_PATH = _GENERATED_AGENT_DIR / "generatedToolCatalog.json"
_TOOL_SCHEMA_PATH = _GENERATED_AGENT_DIR / "generatedToolSchema.json"


@lru_cache(maxsize=1)
def load_generated_tool_catalog() -> list[dict[str, Any]]:
    with _TOOL_CATALOG_PATH.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)

    if not isinstance(payload, list):
        raise ValueError("Generated tool catalog must be a list.")

    return payload


@lru_cache(maxsize=1)
def load_generated_tool_schema() -> dict[str, Any]:
    with _TOOL_SCHEMA_PATH.open("r", encoding="utf-8") as handle:
        payload = json.load(handle)

    if not isinstance(payload, dict):
        raise ValueError("Generated tool schema must be an object.")

    return payload


@lru_cache(maxsize=1)
def build_responses_tool_definitions() -> list[dict[str, Any]]:
    catalog = load_generated_tool_catalog()
    schema = load_generated_tool_schema()
    definitions = schema.get("definitions")
    if not isinstance(definitions, dict):
        raise ValueError("Generated tool schema is missing definitions.")

    tool_definitions: list[dict[str, Any]] = []
    for entry in catalog:
        if not isinstance(entry, dict):
            raise ValueError("Generated tool catalog entries must be objects.")

        tool_name = entry.get("toolName")
        input_type = entry.get("inputType")
        description = entry.get("description")
        strict = entry.get("strict")

        if not isinstance(tool_name, str) or not isinstance(input_type, str):
            raise ValueError("Generated tool catalog entry is missing names.")

        if not isinstance(description, str):
            raise ValueError(
                "Generated tool catalog entry is missing a description."
            )

        if not isinstance(strict, bool):
            raise ValueError(
                "Generated tool catalog entry is missing a strict flag."
            )

        if input_type not in definitions:
            raise ValueError(
                "Generated tool schema is missing definition for " + input_type
            )

        tool_definitions.append(
            {
                "type": "function",
                "name": tool_name,
                "description": description,
                "parameters": _project_schema(
                    definitions[input_type], definitions
                ),
                "strict": strict,
            }
        )

    return tool_definitions


def _project_schema(
    schema: Any,
    definitions: dict[str, Any],
    excluded_definition_names: set[str] | None = None,
    visited: set[str] | None = None,
) -> Any:
    if excluded_definition_names is None:
        excluded_definition_names = {"AgentIntentProgramStep"}

    if visited is None:
        visited = set()

    if schema is None or not isinstance(schema, (dict, list)):
        return schema

    if isinstance(schema, list):
        return [
            _project_schema(item, definitions, excluded_definition_names, visited)
            for item in schema
        ]

    ref = schema.get("$ref")
    if isinstance(ref, str):
        ref_name = ref.replace("#/definitions/", "")
        if ref_name in excluded_definition_names or ref_name in visited:
            return {"type": "object"}

        ref_schema = definitions.get(ref_name)
        if not isinstance(ref_schema, dict):
            return {"type": "object"}

        visited.add(ref_name)
        projected = _project_schema(
            ref_schema, definitions, excluded_definition_names, visited
        )
        visited.remove(ref_name)
        return projected

    projected: dict[str, Any] = {}
    for key, value in schema.items():
        if key in {"definitions", "$schema"}:
            continue

        if key == "properties" and isinstance(value, dict):
            projected[key] = {
                property_name: _project_schema(
                    property_schema,
                    definitions,
                    excluded_definition_names,
                    visited,
                )
                for property_name, property_schema in value.items()
            }
            continue

        if key in {"items", "not", "if", "then", "else"}:
            projected[key] = _project_schema(
                value, definitions, excluded_definition_names, visited
            )
            continue

        if key in {"anyOf", "allOf", "oneOf"} and isinstance(value, list):
            projected[key] = [
                _project_schema(item, definitions, excluded_definition_names, visited)
                for item in value
            ]
            continue

        if key == "additionalProperties" and isinstance(value, dict):
            projected[key] = _project_schema(
                value, definitions, excluded_definition_names, visited
            )
            continue

        projected[key] = value

    if projected.get("type") == "object" and "properties" not in projected:
        projected["properties"] = {}

    return projected
